### ISSUE TITLE
deep image retrieval loss

### ISSUE THREAD

**Issue by chullhwan-song** [chullhwan-song/Reading-Paper#181]
* 잘 이해가 안가는 부분/헷갈리는 부분들이 있어, 여기저기 다양한 논문/Post에서 이해하려고 가져왔음.  

# 목표
* 학습의 최종 목표는 다른 클래스끼의 거리는 크고(inter class는 크고), 같은 클래스안에서의 거리는 (inter class) 작아(작게 뭉치는 모)지는 형태의 모양의 형태가 되어야한다.

## softmax 

![image](https://user-images.githubusercontent.com/40360823/62850109-ed54c380-bd1c-11e9-8917-5246e2f692cb.png)
![image](https://user-images.githubusercontent.com/40360823/62850131-02315700-bd1d-11e9-8234-ceae8ad136bb.png)

* 바로 위의 그림, > separable feature
    * 바로 위의 그림은 inter 클래스는 작고, intra 가 클때의 그림.
         * 이 의미는 클래스 별로 잘 구분되는 것이 않고(inter 클래스는 작고),  클래스 내부에 서로 뭉쳐지지 않는다는 의미( intra 가 클때)
         * softmax가 항상 그렇다는 것은 아니고, 실제로 arcface를 설명하기 위해 비교, 그니까..arcface보다는 inter class가 작고, intra 클래스가 크다는 의미로 해석해야한다고 봄.
         * 또하나의 유사 연구에서는 보통 softmax에 **center loss**를 adding하여 "discriminative feature"형태로 학습..하는 유명한 연구가 있음.

### refer
* https://medium.com/@mathanrajsharma/softmax-activation-function-e582ea53ada7
   * score가 음수가 나오고 , sum =0 이면 문제 발생,  그래서, > exp(e의 자승) 형태로...

## center loss
* 보통 softmax와 결합하여 사용
* 각 클래스의 centroid를 구하여 계산.
![image](https://user-images.githubusercontent.com/40360823/62917952-2eaaa900-bdd9-11e9-855d-8479d1aed8be.png)

## sphere loss
* softmax 기반
* 좀더 자세한 내용은 밑의 arcface
* angle을 이용. > Angular softmax loss
* ? a-softmax와 차이가..같은건가?? 헷갈림.. 
   * sphere loss & arcface & sphere loss는 거의 같은 개념인데..
   * sphere loss & arcface 는 특히 거의 동일하지 않을까? 
   * Geometric Difference 가 차이가 있다고 함. arcface가 가장좋은 이유는  Geometric의 속성이 다른 loss보다 좀더 좋아서...그렇다고 함.ㅠ
![image](https://user-images.githubusercontent.com/40360823/62909414-7f110f00-bdb7-11e9-9d05-acf0d70ce30d.png)
* 클래스간의 상대적 angle에 대한 비교를 통해 그 차이를 이용하여 구분.
* 좀더 자세한 내용은 밑의 arcface

### refer
* https://tmaxai.github.io/post/review_SphereFace/

## Large Margin Cosine Loss : CosFace
* softmax
* 좀더 자세한 내용은 밑의 arcface
* Angular softmax loss(=sphere loss, a-softmax) vs  Large Margin Cosine Loss  (=CosFace, LMCL)
   * 매우 유사.
   *  CosFaces(LMCL)는 이미 정해진 default 크기의 margin m 을 모든 클래스가 그 크기만큼 반드시 차이가 나도록 학습.
        * 이는 sphere loss와 거의 유사한데 이점이 다름.
        * decision boundary 표 비교(밑에서..)
   * a-softmax는 weight만 normalize > feature의 유클리디안 norm(크기)과 cos(θ)를 같이 학습
   * CosFaces는 weight와 feature 모두 normalize > (norm(크기)는 제외됨.) > 클래스 분류에서 theta만..학습에 집중 > 그래서 이게 더 좋다고 나옴.
* angular 기반의 loss는 쓰면서도 너무 헷갈린다..

## arcface 
* softmax
   * arcface는 metric learning이 아니라, sofmax의 업그레이드 버전.
* 거리계산(여기서는 triplet loss가 비교대상)보다는 angle(cosine distnace) 로.. > a-softmax
* 링크 포스트 참조 : [링크](https://github.com/ChoSungMan/awesome-face-recognition/tree/master/paper_review)
    1. 기존 softmax 수식 >  inter-class 간의 차이를 늘려주고, intra-class 끼리 차이를 줄여주는 부분에 대 해서는 최적화 (이는 이 포스트에서 많이 설명..!!)
![image](https://user-images.githubusercontent.com/40360823/67731066-4853a800-fa3a-11e9-8ed1-63dc172c6e75.png)
         * N:batch_size, n:클래스         
     2. 이를 위해, 
![image](https://user-images.githubusercontent.com/40360823/67731220-d16adf00-fa3a-11e9-8776-c39d9513b8ae.png) 
* a-softmax기반..
    * arcface는 cosface에 더 유사한듯..밑의 그림에서 보듯..w와 f를 normalize하니..
* Angular Margin Loss를 이용, 다만 앞에서 언급했듯이..margin plus noramlized방법등이 살짝 비슷한 계열의 loss끼리 다른듯~
![image](https://user-images.githubusercontent.com/40360823/62911472-745a7800-bdbf-11e9-9577-523d96d6328a.png)
![image](https://user-images.githubusercontent.com/40360823/62851256-75889800-bd20-11e9-91d5-b6d315b1f315.png)
* 이미 언급했지만 위의 오른쪽 그림은 Geometric Difference 가 차이가 있다고 함. arcface가 가장좋은 이유는  Geometric의 속성이 다른 loss보다 좀더 좋아서...그렇다고 함.ㅠ
* separable feature보다는 discriminative feature 형태로 해야한다..
    * 바로 아래 왼쪽 discriminative feature 그림은 inter 클래스는 크고(다른 클래스끼리의 거리는 커야한다는 의미) , intra 클래스는 작을때(같은 클래스끼리의 거리가 작아야한다는 의미) 의 그림.(이를 목표로해야함. > 모든 분류의 목적은 이것!!!,)
         * 이 의미는 클래스 별로 구분(inter 클래스는 크고),  클래스 내부에 서로 뭉쳐진다는 의미( intra 가 작다..)
![image](https://user-images.githubusercontent.com/40360823/62850876-550c0e00-bd1f-11e9-8288-b1922dfaea5b.png)
* arcface학습후, triplet loss로 finetuning하면 더 좋다.
    * softmax 학습후, triplet loss로 finetuning하면 더 좋아지는것처럼...
* Additive Angular Margin Loss   
![image](https://user-images.githubusercontent.com/40360823/62908434-027c3180-bdb3-11e9-8fd1-13462b0d7d44.png)
![image](https://user-images.githubusercontent.com/40360823/62908475-36efed80-bdb3-11e9-994a-40269439ccf0.png)
 *  decision boundary 비교
![image](https://user-images.githubusercontent.com/40360823/62908998-b2529e80-bdb5-11e9-9e55-06fe050f4e64.png)
![image](https://user-images.githubusercontent.com/40360823/62915949-49791f80-bdd1-11e9-997f-e2e5cca6a2e4.png)
      * 같은 크래스 angle에 m을 곱해주면 더 가깝다고 함. 즉, 같은 클래스는 당연히 가까워지도록하는것이..기존 개념과 동일. 
     * 다른 클래스에 비해 각도 theta보다... m배 작아지도록 W를 학습.
     * softmax의 boundary 계산
        * 두 클래스(binary classification)에서 a클래스의 W_a, bias b_a와 b클래스 W_b, bias b_b의 parameter가 있고(wx+b),  주어진 score(feature) x가 있을때, boundary 는 Wx+b에서, (W_a-W_b)x+(b_a-b_b) = 0 형태이다.
            * P_a = (W_a+b_a)/((W_a+b_a)+(W_b+b_b))
            * P_b = (W_b+b_b)/((W_a+b_a)+(W_b+b_b))
      * 반면에, 이러한 거리 계산이 아닌 Angle걔념으로 converting!!! 따라서,  반지름이 1인 구 형태로 Mapping 하는 개념이므로, W의 크기(norm) ||W||의 1, bias는 구(원) 중심으로 매핑되므로 0이다.
     * 그래서, base angle 차이식은 ||x||(cos(theta_a)-cos(theta_b))
         *  이때, 당연히,  ||W||, ||b|| 가 normalize의 형태로 와야한다는 의미가 됨.
         * 이때 중요한 의미는 x가 W_a와 W_b중에서 가까운 정도는 angle 즉, theta에 의해 결정된다.
     *  그래서, angle에 대한,  P_a에대한 확률은 ||x||cos(theta_a), P_b=|x||cos(theta_b)
     * softmax 기본 확률 공식으로 대입해보면, 
            * P_a = |x||cos(theta_a)/(|x||cos(theta_a)+||x||cos(theta_b))
            * P_a = |x||cos(theta_b)/(|x||cos(theta_a)+||x||cos(theta_b))
       * 이같은 개념으로, .... normalize하고.등등하면 angle기반의 softmax가 된다.  = modified softmax loss(NSL)
![image](https://user-images.githubusercontent.com/40360823/62915570-d8853800-bdcf-11e9-9506-021697916fc8.png)
            * 이 공식에서 cos theta의 theta값에 margin m을 더해주고나 곱해주는 형태로 진화하는듯하다.
            *  예를들어, Angular-softamx loss(a-softmax)에서, 위의 decision boundary 비교표에서 보면, 같은 클래스에 theta에 m을 곱해줘서 angle이 가까워지도록 유도.. 
            * margin 개념을 도입하면,  modified softmax loss vs a-softmax차이가 다음과 같은 ...모양새가된다고 함. 즉, modified softmax loss에서 margin을 도합한 형태가 a-softmax(sphereface, arcface..)
![image](https://user-images.githubusercontent.com/40360823/62915682-56494380-bdd0-11e9-8554-369b16ea47bc.png)
            * 여기서, a-softmax에서, sphere(a-softmax의 원형인듯..) 와 arcface/cosface와 구분하면,
![image](https://user-images.githubusercontent.com/40360823/62915838-d8d20300-bdd0-11e9-891b-a32697c62aed.png)
            * 다시 여기서,  arcface vs cosface의 형태를 더 구분하려면, 
![image](https://user-images.githubusercontent.com/40360823/62915855-f2734a80-bdd0-11e9-82d5-4d292870c90f.png)

### refer
* https://www.slideshare.net/healess/face-recognitionarcface
* https://slideplayer.com/slide/15962912/
* https://tmaxai.github.io/post/review_SphereFace/

## Siamese loss 
* metric loss
* Contrastive Loss
![image](https://user-images.githubusercontent.com/40360823/62917927-1e92c980-bdd9-11e9-8c89-21761887015f.png)

## triplet loss
* metric loss
![image](https://user-images.githubusercontent.com/40360823/62908641-078db080-bdb4-11e9-984f-dc7fd5b97a8e.png)

## n-pair loss
* metric loss
![image](https://user-images.githubusercontent.com/40360823/62908618-e0cf7a00-bdb3-11e9-94ca-ac3932ef7ec7.png)

## angular loss : Additive Angular Margin Loss
* metric loss
* 기존 거리기반 loss(triplet, n-pair)들은 inner 클래스간의 거리만 의존됨
* 즉, intra-class에 대한 variance에 대한 scale를 무시했다는 의미와 상통. > 위에서 언급한 intra 클래스안에서는 작아야한다는 의미(작게 뭉쳐야..)와 같은 의미인듯..
* 거리에서 각도로...
   * angluar 기반의 loss(sphere loss..)를 metric learning으로 적용한다는 의미도 된다고 봄.
* angluar  loss의 논문과 겹치지 않나쉽다.

### join a-softmax(sphere loss)
![image](https://user-images.githubusercontent.com/40360823/62907516-98fa2400-bdae-11e9-9f8e-4746cc6793c2.png)
